### PR TITLE
feat!: Add overlap to layout

### PR DESCRIPTION
**Pull request by kdheepak**
This PR adds a new feature for the existing `Layout::spacing` method, and introducing a `Spacing` enum.

Now `Layout::spacing` is generic and can take

- zero or positive numbers, e.g. `Layout::spacing(1)` (current functionality)
- negative number, e.g. `Layout::spacing(-1)` (new)
- variant of the `Spacing` (new)

This allows creating layouts with a shared pixel for segments. When `spacing(negative_value)` is used, spacing is ignored and all segments will be adjacent and have pixels overlapping. `spacing(zero_or_positive_value)` behaves the same as before. These are internally converted to `Spacing::Overlap` or `Spacing::Space`. 

Here's an example output to illustrate the layout solve from this PR:

```rust
#[test]
fn test_layout() {
    use crate::layout::Constraint::*;
    let mut terminal = crate::Terminal::new(crate::backend::TestBackend::new(50, 4)).unwrap();
    terminal
        .draw(|frame| {
            let [upper, lower] = Layout::vertical([Fill(1), Fill(1)]).areas(frame.area());

            let (segments, spacers) = Layout::horizontal([Length(10), Length(10), Length(10)])
                .flex(Flex::Center)
                .split_with_spacers(upper);

            for segment in segments.iter() {
                frame.render_widget(
                    crate::widgets::Block::bordered()
                        .border_set(crate::symbols::border::DOUBLE),
                    *segment,
                );
            }
            for spacer in spacers.iter() {
                frame.render_widget(crate::widgets::Block::bordered(), *spacer);
            }

            let (segments, spacers) = Layout::horizontal([Length(10), Length(10), Length(10)])
                .flex(Flex::Center)
                .spacing(-1) // new feature
                .split_with_spacers(lower);

            for segment in segments.iter() {
                frame.render_widget(
                    crate::widgets::Block::bordered()
                        .border_set(crate::symbols::border::DOUBLE),
                    *segment,
                );
            }
            for spacer in spacers.iter() {
                frame.render_widget(crate::widgets::Block::bordered(), *spacer);
            }
        })
        .unwrap();
    dbg!(terminal.backend());
}
```

```plain
┌────────┐╔════════╗╔════════╗╔════════╗┌────────┐
└────────┘╚════════╝╚════════╝╚════════╝└────────┘
┌─────────┐╔════════╔════════╔════════╗┌─────────┐
└─────────┘╚════════╚════════╚════════╝└─────────┘
```

Currently drawing a border on top of an existing border overwrites it. Future PRs will allow for making the border drawing handle overlaps better.

### CODE REVIEW

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:2293]: Try to use unique (distinct) test names here to make it possible to work out which test fails easily. The purpose of ::test_name is to change the test name from `case_01`, ... to `case_01_length_overlap.

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:2307]: Not related to this specific test, but the output of these tests when they fail is particularly poor. Would be nice to have a small struct for this that has start and width instead of the tuple.

E.g.
```
thread 'layout::layout::tests::split::fill_overlap::case_07_flex10' panicked at src/layout/layout.rs:2488:13:
assertion failed: `(left == right)`

Diff < left / right > :
 [
     (
         0,
<        55,
>        50,
     ),
     (
<        45,
<        55,
>        50,
>        50,
     ),
 ]
```

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:2309]: Any idea why these are backwards compared to the rest of the lib (actual, expected)?

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:689]: Would be nice if this didn't have to be a special case. I'm surprised that it does have to be. What's different about having an overlap of 1 and an overlap of 0?

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:2328]: I agree, the tests failing here are hard to debug. What about something like this?

image

https://github.com/ratatui/ratatui/pull/1398/commits/03be41a02bee5ff0be386b4f4003aa94e60f4bbc

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: Done in https://github.com/ratatui/ratatui/pull/1398/commits/eff582110be19226c57e5778590cc7a99018e1aa

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: There's a number of places in this file where it is backwards. I copy pasted another test from this file and edited. This is something that we should fix in a separate PR.

Comment (kdheepak) [reply src/layout/layout.rs]: Opened an issue to track this: https://github.com/ratatui/ratatui/issues/1399

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:689]: overlap 1 only adds constraints for segments but overlap 0 adds constraints for all variables (i.e. segments and spacers).

File: src/layout/layout.rs
Comment (joshka) [reply src/layout/layout.rs:2329]: Perhaps check starts and widths separately? It's tough to know what to do here (and it's a bit of a micro-optimization I guess)

I can see what you're doing there. The idea is good - perhaps could work by showing the rects as full rects with borders.

File: src/layout/layout.rs
Comment (joshka) [reply src/layout/layout.rs:689]: I think what you're saying here is that the overlap eats up any space that would have been allocated to the spacer, but the spacer doesn't eat any space allocated to the segments. Does that sound correct?

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:2343]: I think regardless of what we do here that should happen in a separate PR, because we have to change every test here to use the new format.

Comment (kdheepak) [reply src/layout/layout.rs:2343]: I've opened an issue to track this: https://github.com/ratatui/ratatui/issues/1400

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:689]: To make sure we are on the same page, here's the terminology.

We have the following area:

```plain
<───────────────────────────────────area_width─────────────────────────────────>
┌─area_start                                                          area_end─┐
v                                                                              v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
```

We create elements `Vec<Element>` where each `Element` has a `start` and `end`:

```plain
┌────┬───────────────────┬────┬───element.start───┬────┬───────────────────┐
│    │                   │    │                   │    │                   │
v    v                   v    v                   v    v                   v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
    ^                   ^    ^                   ^    ^                   ^    ^
    │                   │    │                   │    │                   │    │
    └───────────────────┴────┴───element.end─────┴────┴───────────────────┴────┘
```

"segments" are `Vec<Element>` only for the areas of interest:

```plain
     ┌────────────────────────┬───segment.start────────┐
     │                        │                        │
     v                        v                        v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
                        ^                        ^                        ^
                        │                        │                        │
                        └──────────segment.end───┴────────────────────────┘
```

"spacers" are `Vec<Element>` the remaining gaps:

```plain
┌────────────────────────┬────────spacer.start────┬────────────────────────┐
│                        │                        │                        │
v                        v                        v                        v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
    ^                        ^                        ^                        ^
    │                        │                        │                        │
    └────────────────────────┴────spacer.end──────────┴────────────────────────┘
```

Comment (joshka) [reply src/layout/layout.rs:689]: yup - this aligns with my understanding. But adding this image is good for making it easy to visualize conversations, so thanks. The only thing I'd add to the description is that the variable's value really effectively points to the left side of the cell. This subtle distinction helps make it a bit more clear why the end is 1 more than the cell index, and makes the floating point values less ambiguous.

Comment (kdheepak) [reply src/layout/layout.rs:689]: > I think what you're saying here is that the overlap eats up any space that would have been allocated to the spacer, but the spacer doesn't eat any space allocated to the segments. Does that sound correct?

Previously, we were always adding this constraint:

```rust
    // all variables are in ascending order
    for (&left, &right) in variables.iter().tuple_windows() {
        solver.add_constraint(left | LE(REQUIRED) | right)?;
    }
    Ok(())
```

If we use individual variables:

```plain
┌────┬───────────────────┬────┬─────variables─────┬────┬───────────────────┬────┐
│    │                   │    │                   │    │                   │    │
v    v                   v    v                   v    v                   v    v
┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐┌──────────────────┐┌   ┐
     │     Max(20)      │     │      Max(20)     │     │      Max(20)     │
└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘└──────────────────┘└   ┘
^    ^                   ^    ^                   ^    ^                   ^    ^
└v1  └v2                 └v3  └v4                 └v5  └v6                 └v7  └v8
```

this does:

```
v1 <= v2
v2 <= v3
v3 <= v4
v4 <= v5
v5 <= v6
v6 <= v7
v7 <= v8
```

But when we have overlaps, we cannot do this anymore.

Comment (joshka) [reply src/layout/layout.rs:689]: yes, but we can do generically:

```
v1 - v2 <= 0 // not sure about this one
v2 - v3 <= 0
v3 - v4 <= overlap
v4 - v5 <= 0
v5 - v6 <= overlap
v6 - v7 <= 0
v7 - v8 <= 0 // not sure about this one either
```

Comment (kdheepak) [reply src/layout/layout.rs:689]: I _think_ we have to force `overlap`, i.e.

```
v3 - v4 == overlap
```

instead of 

```
v3 - v4 <= overlap
```

Otherwise we may get overlaps that are not what the user wants.

**Comment by orhun** [comment]
Hmm, I'm curious about where this might be practically useful 🤔

**Comment by joshka** [comment]
> Hmm, I'm curious about where this might be practically useful 🤔

Collapsed borders is the primary use case for this.

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:457]: This makes `foo.spacing(1).overlap(1)` different to `foo.overlap(1).spacing(1)`. Mostly our builder methods should not have order dependent methods like this.

Comment (kdheepak) [reply src/layout/layout.rs:457]: I agree. The only way around this though is by making a breaking change to `spacing`. I'm open to suggestions on what to do here.

Comment (joshka) [reply src/layout/layout.rs:457]: If these are truly not able to be set together, then they likely deserve to be on an enum with mutually exclusive variants. My initial question on this was on the discord thread sort of had that idea in mind (setting spacing to negative values). If we took that approach, then spacing could be migated to accept `Into<Spacing>` => `enum Spacing { Spaced(u16), Overlap(u16) }`, which would give space for documentation. Alternatively spacing could just become i16 (or i32 if you want to allow layouts that have spacers / overlaps that are > 32767). This would be technically breaking, but not in a way that generally matters to anyone.

Comment (kdheepak) [reply src/layout/layout.rs:457]: I think making it `i16` or `i32` would work. It'd break apps that use `spacing` like we do in constraint explorer:

https://github.com/ratatui/ratatui/blob/edcdc8a8147a2f450d2c871b19da6d6383fd5497/examples/constraint-explorer.rs#L47

But maybe that's fine?

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:59]: Why do we need from u32? There shouldn't be any code that already uses u32 for backwards compat right?

Comment (kdheepak) [reply src/layout/layout.rs]: I thought I tried that first. If I don't implement `i32`, I get compile errors on some tests:

image

But it looks like implementing `u32` is not required. Resolved in https://github.com/ratatui/ratatui/pull/1398/commits/3b817e598a49c591ab51b032507a12a9808f1c85

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: Removed `overlap` method in https://github.com/ratatui/ratatui/pull/1398/commits/57dbafe8c7fd42d1901cbc41810b3c609f578ae0

**Comment by kdheepak** [comment]
I made some changes in https://github.com/ratatui/ratatui/pull/1398/commits/ae72e0b2a0b0e787a4e2a796c5a99a5c4d02b737 that change the behavior of overlap in SpaceBetween and SpaceAround. imo, this is more consistent with the behavior of SpaceBetween and SpaceAround with positive spacing, i.e. rules and precedence for positive spacing are the same as negative spacing. 

Apply this patch to `constraint-explorer.rs` to test out this PR:

```
diff --git a/examples/constraint-explorer.rs b/examples/constraint-explorer.rs
index a467551..a729955 100644
--- a/examples/constraint-explorer.rs
+++ b/examples/constraint-explorer.rs
@@ -44,7 +44,7 @@ fn main() -> Result<()> {
 #[derive(Default)]
 struct App {
     mode: AppMode,
-    spacing: u16,
+    spacing: i16,
     constraints: Vec<Constraint>,
     selected_index: usize,
     value: u16,
```

**Comment by joshka** [comment]
@kdheepak can you please update the first comment to cover document this for a changelog entry?

**Comment by orhun** [comment]
Can you wait for my review before merge? I'd like to see how this works :)

**Review by joshka — COMMENTED** [review]
Need to probably add a change to the constraint explorer to support this too

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:176]: My bad, I implemented a custom `Default` and the undid the change to cause this

**Comment by joshka** [comment]
Oh, and btw, I'm +Approve generally on this now - I wasn't 100% sure at first, but switching to the enum makes it work well

File: src/layout/layout.rs
Comment (orhun) [reply src/layout/layout.rs:471]: I think they are now added in the doc comments of `Spacing`.

File: src/layout/layout.rs
Comment (orhun) [inline src/layout/layout.rs:608]: We can actually drop the `-rs` suffix here since the library is called "cassowary".

File: src/layout/layout.rs
Comment (orhun) [inline src/layout/layout.rs:2761]: We should probably do a buffer assertion here :)

**Review by orhun — APPROVED** [review]
LGTM!

What is in your mind about adding a practical example for this? I think we can either use it in the examples somewhere or mention it in the recipes (e.g. collapsing borders seems related to this).

**Comment by joshka** [comment]
> What is in your mind about adding a practical example for this? I think we can either use it in the examples somewhere or mention it in the recipes (e.g. collapsing borders seems related to this).

This is step one towards collapsing borders, Step 2 is making block able to handle overwriting borders - that can be also pretty complex, as it requires merging symbol sets / building up table(s) of combinations and mappings.

This looks something like:
1. app draws initial block
2. app draws second block - for every border cell, the block checks for a non-blank character and then computes the new character to use based on the desired character from the symbol set, combined with the existing symbol in the current cell.

I imagine this would likely involve hard coding a bunch of combinations to start. It's easy if the lines are in the same line set (normal, dotted, heavy, double, ...), but there combinations for each of the intersection of sets. E.g. all the following are different:
```text
┥ 	┦ 	┧ 	┨ 	┩ 	┪ 	┫ 
```

Then on top of that, because we offer custom sets, this should be something that can be configured for the user's particular choice of sets. This part doesn't have to happen in an initial PR though, but it should happen - we can default to just overwriting when we don't have a mapping.

The thread related to this is at https://discord.com/channels/1070692720437383208/1291063903987433503

File: src/layout/layout.rs
Comment (joshka) [inline src/layout/layout.rs:73]: needs to be re-exported in src/layout.rs

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: Resolved in 276694932388

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: Resolved in 590ac83f23f8

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: Resolved in a45d4dbc43ea

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:73]: Resolved in a0308d68eaf6

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs:608]: Made an issue since it might be a good first issue for someone intending to contribute - #1423

File: src/layout/layout.rs
Comment (kdheepak) [reply src/layout/layout.rs]: This was just my function to print the ascii content for the PR description. I've removed the test now.

**Comment by kdheepak** [comment]
> Need to probably add a change to the constraint explorer to support this too

I'd like to do this after we figure out better drawing methods for overlapping.

**Comment by kdheepak** [comment]
This should be good to go! If the PR description looks good, I can merge.

**Comment by joshka** [comment]
I'd suggest cutting the example down to something more concise for the changelog - just two constraints, but otherwise yep - good to merge.

**Comment by kdheepak** [comment]
I wanted to show 3 blocks, because it shows the left and right border for the middle block. I can decrease the width of the output?

**Pull request merged**
